src/library/scala/util/Using.scala

```diff
@@ -0,0 +1,286 @@
+package scala.util
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.util.control.{ControlThrowable, NonFatal}
+
+/** A utility for performing automatic resource management. It can be used to perform an
+  * operation using resources, after which it will close the resources, in reverse order
+  * of their creation. The resource opening, operation, and resource closing are wrapped
+  * in a `Try`.
+  *
+  * If more than one exception is thrown by the operation and closing resources,
+  * the exception thrown ''first'' is returned within the `Try`, with the other exceptions
+  * [[java.lang.Throwable.addSuppressed(Throwable) added as suppressed exceptions]]
+  * to the one thrown first. This is the case ''unless'' a later exception is
+  * [[scala.util.control.NonFatal fatal]], and the one preceding it is not. In that case,
+  * the first exception is added as a suppressed exception to the fatal one, and the fatal
+  * one is thrown. If an exception is a
+  * [[scala.util.control.ControlThrowable ControlThrowable]], no exception will be added to
+  * it as a suppressed exception.
+  *
+  * @example
+  * {{{
+  * val lines: Try[List[String]] = Using(resource1) { r1 =>
+  *   r1.lines.toList
+  * }
+  * }}}
+  *
+  * @example
+  * {{{
+  * val lines: Try[Seq[String]] = for {
+  *   r1 <- Using(resource1)
+  *   r2 <- Using(resource2)
+  *   r3 <- Using(resource3)
+  *   r4 <- Using(resource4)
+  * } yield {
+  *   // use your resources here
+  *   r1.lines ++ r2.lines ++ r3.lines ++ r4.lines
+  * }
+  * }}}
+  */
+final class Using[R](resource: => R) {
```

Comment:
Why is `resource` not a value?

Reply:
I am consider the following snippet, but I'm having trouble convincing myself that having it not by-name will prevent it from handling exceptions during resource creation.

```scala
for {
  foo <- Using(new Foo)
  bar <- Using(Bar(throw new Exception))
} yield foo.withBar(bar)

// equivalent to
Using(new Foo)
  .flatMap(foo => Bar(throw new Exception)
    .map(bar => foo.withBar(bar)))
```

Could you help me out @Ichoran? I believe you suggested they be by-name.

Reply:
@NthPortal - What should the behavior of `Using(new Foo(throw new Exception))` alone be?  (Also, I vaguely recall that I was talking about the multi-argument version.)  If it's by-name, then `Using` can catch the exception.  If not, you still have to remember to catch the exception associated with allocating the resource in the first place (even though it handles all the other exceptions).  The latter, I think, will yield awkward patterns like `Try(Using(new Foo)(f).toTry)).flatten`.

Reply:
@Ichoran good point. I was only thinking about a nested situation, when what's actually important here is when it's the top-level one. Thanks!

I agree that `Using(new Foo(throw new Exception))` should return a `Failure`, not throw an exception.